File: BST/bst.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    struct node* left;
    int data;
    struct node* right;

};

struct node* root = NULL;
struct node* newnode;

void Insertion(int data) {
   struct node *tempNode = (struct node*) malloc(sizeof(struct node));
   struct node *current;
   struct node *parent;

   tempNode->data = data;
   tempNode->left = NULL;
   tempNode->right = NULL;

   //if tree is empty, create root node
   if(root == NULL) {
      root = tempNode;
   } else {
      current = root;
      parent  = NULL;

      while(1) {                
         parent = current;

         //go to left of the tree
         if(data < parent->data) {
            current = current->left;                
            
            //insert to the left
            if(current == NULL) {
               parent->left = tempNode;
               return;
            }
         }
			
         //go to right of the tree
         else {
            current = current->right;
            
            //insert to the right
            if(current == NULL) {
               parent->right = tempNode;
               return;
            }
         }
      }            
   }
}

struct node* Search(int search_element)
{
   struct node *current = root;
   
   while(1)
   {
       if(current != NULL)
       {
           if(current->data == search_element)
                return current;
                
            else if(current->data > search_element)
                current = current->left;
                
            else
                current = current->right;
                
       }
       
       else{     
        current = NULL;
        return current;
       }
   }
  
  
    
    
}

void InOrder_Traversal(struct node* node)  //Ascending order result 
{
     if (node == NULL) 
        return; 
  
    /* first recur on left child */
    if(node->left != NULL)
        InOrder_Traversal(node->left); 
  
    /* then print the data of node */
 
    printf(" %d ->",node->data);
  
    /* now recur on right child */
    if(node->right != NULL)
        InOrder_Traversal(node->right); 
}

void PreOrder_Traversal(struct node *node)
{
    if(node == NULL)
        return;
        
    printf("%d ->",node->data);
        
    if(node->left != NULL)
        PreOrder_Traversal(node->left);
        
    
    if(node->right != NULL)
        PreOrder_Traversal(node->right);
        
    
}

void PostOrder_Traversal(struct node *node)
{
    if(node == NULL)
        return;
        
   
        
    if(node->left != NULL)
        PostOrder_Traversal(node->left);
        
    
    if(node->right != NULL)
        PostOrder_Traversal(node->right);
        
     printf("%d ->",node->data);
}

void main()
{

    /* Sample input */
    
    Insertion(50);
    Insertion(49);
    Insertion(52);
    Insertion(46);
    Insertion(48);
    Insertion(53);
    Insertion(57);
    Insertion(45);

    InOrder_Traversal(root);
    
}

/* Sample Output */

/*

 45 -> 46 -> 48 -> 49 -> 50 -> 52 -> 53 -> 57 ->  
 
 */

File: BST/bst.py
class Node:
    def __init__(self,data):
        
        self.left = None
        self.data = data
        self.right = None
        
    
    def insert(self,data):
        
        """
        Method to insert data into Binary Tree
        """
        
    # If root node already present    
        if self.data:
            if data < self.data:
               if self.left is None:
                   self.left = Node(data)
            
               else:
                   self.left.insert(data)
                   
            
            elif data > self.data :
                if self.right is None:
                    self.right = Node(data)
                
                else:
                    self.right.insert(data)
    
    # If root not present
        else:
            self.data = data
            
    
    def printInOrderTree(self):
        """
        Method to print the tree data In Order
        """
        if self.left:
            self.left.printInOrderTree()
            
        print(self.data)
        
        if self.right:
            self.right.printInOrderTree()
            
    def printPreOrderTree(self):
        """
        Method to print the tree data Pre Order
        """
        print(self.data)
        
        if self.left:
            self.left.printPreOrderTree()
        
        if self.right:
            self.right.printPreOrderTree()
            
    def printPostOrderTree(self):
        """
        Method to print the tree data Post Order
        """
        if self.left:
            self.left.printPostOrderTree()
        
        if self.right:
            self.right.printPostOrderTree()
            
        print(self.data)
        

BST = Node(4)
BST.insert(2)
BST.insert(1)
BST.insert(3)
BST.insert(5)
BST.insert(6)

print("\nInorder Tree : ")
BST.printInOrderTree()

print("\nPre Order Tree : ")
BST.printPreOrderTree()

print("\nPost OrderTree Tree : ")
BST.printPostOrderTree()

# Output 

#Inorder Tree : 
#1
#2
#3
#4
#5
#6

#Pre Order Tree : 
#4
#2
#1
#3
#5
#6

#Post OrderTree Tree : 
#1
#3
#2
#6
#5
#4

